### PR TITLE
Add ComponentInstaller SPI and use it for OpenTelemetry SDK

### PR DESCRIPTION
**Pull request by pavolloffay**
Supersedes #1619 

@trask @iNikem  opening draft for now as it needs some cleanup. Let me know if it loos good and I will clean it up.

### CODE REVIEW

**Comment by pavolloffay** [comment]
@trask your comment from #1619 

> While most of the work should be in after hook, because we want bytebuddy agent installed as early as possible so it doesn't need to retransform (many) classes (which has performance impact, due to the retransform itself, as well as due to not being able to use FieldBackedProvider)

When are actually the instrumentations applied to the classes? Is it on https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/InstrumentationModule.java#L111-L130 or `AgentBuilder.installOn` https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java#L160 (I thought it is on the `installOn`)

**Comment by trask** [comment]
> When are actually the instrumentations applied to the classes? (I thought it is on the `installOn`)

ya, inside of `installOn` it will iterate over the already loaded classes, and apply instrumentation (retransform) as needed

File: javaagent-bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/AgentInitializer.java [hunk 1]
```diff
@@ -247,7 +144,7 @@ private static boolean isDebugMode() {
    *
    * @return true if we detect a custom log manager being used.
    */
-  private static boolean isAppUsingCustomLogManager() {
+  public static boolean isAppUsingCustomLogManager() {
```
Comment (trask) [inline javaagent-bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/AgentInitializer.java:147]: can this method be moved to `AgentInstaller` also?

**Review by trask — COMMENTED** [review]
It's looking good to me.

I like the moving of code from AgentInitializer (bootstrap class loader) to AgentInstaller (agent class loader), great to get out of the bootstrap class loader and do as much as possible in the agent class loader 👍

**Comment by pavolloffay** [comment]
There is smth wrong tests are failing but I am able to run the agent on my test app.

**Comment by pavolloffay** [comment]
will be done in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1848

**Pull request closed without merging**